Migrate and revise environment setup, update README index (#58)
datamade/how-to: environment-setup.md

# 💻 Setting up your development environment

There are many guides to setting up your computer for development. Each has its own merits. At DataMade, we perform a mix of maintaining legacy projects and developing new projects, often in parallel. Our toolkit is therefore optimized for managing many, isolated versions of packages and getting up and running on new projects quickly.

#### Contents

- [Version control](#version-control)
- [Text editor](#text-editor)
- [Python](#python)
- [Docker](#docker)
- [Security](#security)
- [Data](#data)
- [Geospatial data](#geospatial-data)
- [Static sites](#static-sites)

> For Linux users, most of these tools are available as aptitude packages. When possible, prefer binary packages to building from source. For Mac users, most of these tools are available as [Homebrew](http://brew.sh/) packages. Homebrew is by far the easiest way to manage packages on OS X.

Reviewer: This guide currently privileges Linux, but MacOS is now the de facto standard for DataMade. (Also, not all Linux distros use aptitude as their package manager.) I think we can comfortably remove the Linux references, since you have to know how to install packages to run Linux anyway.

```suggestion
When possible, we prefer installing binaries from a package manager to building packages from source. For Mac users, most of these tools are available as [Homebrew](http://brew.sh/) packages. Homebrew is by far the easiest way to manage packages on MacOS.
```

## Version control

> We use GitHub and Git to keep our work under version control. Note that we prefer the `git` CLI to the GitHub desktop GUI.

Reviewer: It might be helpful to note that git is usually built-in:

```suggestion
We use GitHub and Git to keep our work under version control. Note that we prefer the `git` CLI to the GitHub desktop GUI. The Git CLI is built-in on MacOS and most Linux distributions, usually with an acceptably recent version; to check whether you already have Git installed, run `which git`.
```

#### Packages

* [Git](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git)

## Text editor

> Most of us use Sublime as our text editor of choice. If you do, too, then you will need to make some modifications to your settings.

Reviewer: Nitpick, but I don't think this sentence is true anymore. More accurate might be:

```suggestion
DataMade developers use a variety of different text editors, and we leave it up to you to decide which one you prefer to use. However, there are two configurations that we encourage you to make no matter which editor you use:

 1. Automatically trim trailing whitespace
 2. Set default tab size for Python files to 4, and set default HTML/JavaScript tab size to 2
 
Historically, Sublime Text has been the most widely-used editor at DataMade, so the following instructions will demonstrate how to set these settings for Sublime Text.
```

In the nav, under "Sublime Text," you will find a "Preferences" fly open, where you will see "Settings - User." Here, you can override the default settings (i.e., do not make changes to the "Settings - Default" file). In the User file, add two things:

```json
    "trim_trailing_white_space_on_save": false,
    "tab_size": 4,
```

‼ **Note:** You are welcome to explore other text editor options, e.g., Atom, Vim, etc.

#### Packages

* [Sublime Text 3](https://www.sublimetext.com/3)

## Python

> At DataMade, you'll run most Python processes in containers. However, it's still useful to have a fresh install of Python on your machine to keep your system Python isolated (it's important!) and to use a later version of Python (if you're on a Mac, your system Python is probably version 2.7).

Reviewer: What is the "(it's important!)" parenthetical doing for us here?

Author: Vaguely underscoring why you might want to keep your system's Python isolated.

> When you aren't using containers, DataMade recommends you conduct Python work in virtual environments (virtualenvs). Virtualenvs help enforce dependency separation between your projects and make it a lot easier for other users to replicate your work on their computers. The Python ecosystem contains a lot of options for managing your environments, from the built-in `virtualenv` package to bundled package and environment management with `conda`. As traditionalists, we like `virtualenvwrapper`, which provides a few convenience functions you can use from your terminal to create, activate, deactivate, and remove virtual environments.

Reviewer: I don't really think our preference for virtualenvwrapper is traditional:

```suggestion
When you aren't using containers, DataMade recommends you conduct Python work in virtual environments (virtualenvs). Virtualenvs help enforce dependency separation between your projects and make it a lot easier for other users to replicate your work on their computers. The Python ecosystem contains a lot of options for managing your environments, from the built-in `virtualenv` package to bundled package and environment management with `conda`. We like `virtualenvwrapper`, which provides a few convenience functions you can use from your terminal to create, activate, deactivate, and remove virtual environments.
```

* [virtualenv](https://virtualenv.pypa.io/en/stable/)
* [virtualenvwrapper](https://virtualenvwrapper.readthedocs.io/en/latest/install.html#basic-installation)
    * virtualenvwrapper contains useful shortcut commands for all stages of virtualenv use. [Install it](https://virtualenvwrapper.readthedocs.io/en/latest/install.html#basic-installation), making sure to follow the instructions for editing your shell startup file.
    * If you run into ownership errors and you installed Python via Homebrew, add `VIRTUALENVWRAPPER_PYTHON=/usr/local/bin/python3` to your shell startup file prior to the lines you added for virtualenvwrapper, and try again.

## Docker

Containers are a popular, modern approach to packaging and running software. We use the Docker engine to create, run, and destroy containers for our applications during local development. This makes it infinitely easier to manage dependencies across 5+ years of web applications.

#### Packages

* [Docker Community Edition](https://www.docker.com/products/container-runtime)

>      * A popular and feature-rich container engine. [Mac users, look no further for installation instructions](https://docs.docker.com/v17.09/docker-for-mac/install/). Unix users, open the Mac instructions and find your distribution in the sidebar.

Reviewer: Nitpick:

```suggestion
     * A popular and feature-rich container engine. [Mac users, look no further for installation instructions](https://docs.docker.com/v17.09/docker-for-mac/install/). Linux users, open the Mac instructions and find your distribution in the sidebar.
```

## Security

Cryptographic security is essential for developers. You don't have to know the guts of how these tools work, but you should have them installed and get comfortable using them.

To enable hard drive encryption on a Macbook, go to System Preferences > Security & Privacy > FileVault. Turn on FileVault and save a recovery key in a safe place, such as LastPass. The process will take some time, probably about a day depending on what you're doing and what model computer you have. It can run in the background as long as your computer is awake and connected to power.

#### Packages

* [SSH](http://linuxcommand.org/man_pages/ssh1.html)
     * Secure Shell, a protocol for communicating securely over unsecured networks. We use it to push and pull from Git remotes and to access our servers. It comes pre-installed as a command line tool on all Mac and Linux distributions; [generate an SSH key](https://help.github.com/articles/generating-an-ssh-key/) if you haven't already and follow the instructions for adding it to your GitHub account.
* [GPG](https://www.gnupg.org/)

>      * GNU Privacy Guard, a command line tool for encrypting and decrypting files. Mac users can `brew install gnupg2`, while enthusiasts of the "subtler arts of computing" can [compile it](https://www.gnupg.org/download/index.en.html). Then, [configure your key](https://keyring.debian.org/creating-key.html) by hand.

Reviewer: The "subtler arts of computing" joke is cute but I think it's not very clear, and goes against our instructions about preferring package managers above:

```suggestion
     * GNU Privacy Guard, a command line tool for encrypting and decrypting files. Mac users can `brew install gnupg2`. Then, [configure your key](https://keyring.debian.org/creating-key.html) by hand.
```

* [Blackbox](https://github.com/StackExchange/blackbox)
     * StackExchange's open-source CLI for keeping secrets secure under public version control. Follow [our excellent guide](https://github.com/datamade/deploy-a-site/blob/master/Setup-blackbox.md) (internal link) to use it.

## Data

We try to maintain a consistent and standardized toolkit for all of our data work. We know that there are many good options for working with data, and we are always open to hearing arguments for new additions to this toolkit. But these tools have stood the test of time, and you'll see them crop up over and over in DataMade's work.

🚨  **Note:** If you're working on a new DataMade application, most of these dependencies (e.g., Postgres) should be containerized, i.e., you don't need to install them directly on your computer, and you can skip this section. If you're working on a legacy application that does not include containerization artifacts, read on for our installation tips.

#### Packages

> * Bash/Unix

Reviewer: I think we mean UNIX tools, not the operating system paradigm:

```suggestion
* Bash and basic UNIX tools
```
